CovariantEqualsCheck: covariant equals on top-level method not flagged in compact source files (JEP 512)

Parent Issue #19971

I have read check documentation: https://checkstyle.org/checks/coding/covariantequals.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > config.xml <<'EOF'
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="CovariantEquals"/>
  </module>
</module>
EOF

vivek@Viveks-MacBook-Air checkstyle % cat > InputCovariantEqualsCompact.java <<'EOF'
public boolean equals(String other) {
    return false;
}

void main() { }
EOF

vivek@Viveks-MacBook-Air checkstyle % RUN_LOCALE="-Duser.language=en -Duser.country=US"
java $RUN_LOCALE -jar target/checkstyle-13.8.0-SNAPSHOT-all.jar -c config.xml InputCovariantEqualsCompact.java

Starting audit...
Audit done.
```

---

**Describe what you expect in detail.**

CovariantEqualsCheck must flag a covariant `equals` method declared at the top level of a compact source file. Under JEP 512 the implicit class is a real class that inherits from `Object`, so a `boolean equals(String)` declared next to no `boolean equals(Object)` override is exactly the design mistake the check exists to catch. The same code inside an ordinary class is flagged with `covariant equals without overriding equals(java.lang.Object).`

Expected on the compact input: one violation on line 1: `covariant equals without overriding equals(java.lang.Object). [CovariantEquals]`.

Root cause: `CovariantEqualsCheck` registers for `CLASS_DEF`, `ENUM_DEF`, `RECORD_DEF`, `LITERAL_NEW` and walks each one's body to find `equals` methods. It does not register for `COMPACT_COMPILATION_UNIT`. In a compact source file the top-level methods are members of the implicit class, but the check never enters an implicit-class context because there is no `CLASS_DEF` for it. The `equals(String)` is silently ignored.

Fix: also register for `COMPACT_COMPILATION_UNIT` and apply the same body-walk logic to its top-level `METHOD_DEF` children.

**Same code inside an ordinary class:**

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > InputCovariantEqualsOrdinary.java <<'EOF'
public class InputCovariantEqualsOrdinary {
    public boolean equals(String other) {
        return false;
    }

    public void main() { }
}
EOF

vivek@Viveks-MacBook-Air checkstyle % java $RUN_LOCALE -jar target/checkstyle-13.8.0-SNAPSHOT-all.jar -c config.xml InputCovariantEqualsOrdinary.java

Starting audit...
[ERROR] InputCovariantEqualsOrdinary.java:2:20: covariant equals without overriding equals(java.lang.Object). [CovariantEquals]
Audit done.
Checkstyle ends with 1 errors.
```

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/CovariantEqualsCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/CovariantEqualsCheck.java
@@ -49,6 +49,16 @@ import com.puppycrawl.tools.checkstyle.utils.CheckUtil;
  * </p>
  *
  * <p>
+ * Note: Compact source files
+ * (<a href="https://openjdk.org/jeps/512">JEP 512</a>)
+ * are skipped by design. Implicit classes in compact source files are not
+ * reusable types and cannot be referenced by name, so they cannot participate
+ * in the polymorphic contexts and collections where covariant {@code equals()}
+ * silently falls back to identity comparison. The rationale for this check
+ * does not extend to compact source files.
+ * </p>
+ *
+ * <p>
  * Inspired by <a href="https://www.cs.jhu.edu/~daveho/pubs/oopsla2004.pdf">
  * Finding Bugs is Easy, chapter '4.5 Bad Covariant Definition of Equals (Eq)'</a>:
  * </p>

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/coding/CovariantEqualsCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/coding/CovariantEqualsCheck.xml
@@ -23,6 +23,16 @@
  &lt;/p&gt;
 
  &lt;p&gt;
+ Note: Compact source files
+ (&lt;a href="https://openjdk.org/jeps/512"&gt;JEP 512&lt;/a&gt;)
+ are skipped by design. Implicit classes in compact source files are not
+ reusable types and cannot be referenced by name, so they cannot participate
+ in the polymorphic contexts and collections where covariant &lt;code&gt;equals()&lt;/code&gt;
+ silently falls back to identity comparison. The rationale for this check
+ does not extend to compact source files.
+ &lt;/p&gt;
+
+ &lt;p&gt;
  Inspired by &lt;a href="https://www.cs.jhu.edu/~daveho/pubs/oopsla2004.pdf"&gt;
  Finding Bugs is Easy, chapter '4.5 Bad Covariant Definition of Equals (Eq)'&lt;/a&gt;:
  &lt;/p&gt;

--- a/src/site/xdoc/checks/coding/covariantequals.xml
+++ b/src/site/xdoc/checks/coding/covariantequals.xml
@@ -28,6 +28,16 @@
         </p>
 
         <p>
+          Note: Compact source files
+          (<a href="https://openjdk.org/jeps/512">JEP 512</a>)
+          are skipped by design. Implicit classes in compact source files are not
+          reusable types and cannot be referenced by name, so they cannot participate
+          in the polymorphic contexts and collections where covariant <code>equals()</code>
+          silently falls back to identity comparison. The rationale for this check
+          does not extend to compact source files.
+        </p>
+
+        <p>
           Inspired by <a href="https://www.cs.jhu.edu/~daveho/pubs/oopsla2004.pdf">
           Finding Bugs is Easy, chapter '4.5 Bad Covariant Definition of Equals (Eq)'</a>:
         </p>

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/CovariantEqualsCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/CovariantEqualsCheckTest.java
@@ -25,6 +25,7 @@ import static com.puppycrawl.tools.checkstyle.checks.coding.CovariantEqualsCheck
 import org.junit.jupiter.api.Test;
 
 import com.puppycrawl.tools.checkstyle.AbstractModuleTestSupport;
+import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
 
 public class CovariantEqualsCheckTest
     extends AbstractModuleTestSupport {
@@ -81,6 +82,36 @@ public class CovariantEqualsCheckTest
         assertWithMessage("Required tokens should not be null")
                 .that(check.getRequiredTokens())
                 .isNotNull();
+    }
+
+    @Test
+    public void testCovariantEqualsCompactSourceFile()
+            throws Exception {
+        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        verifyWithInlineConfigParser(
+                getNonCompilablePath(
+                    "InputCovariantEqualsCompactSourceFile.java"),
+                expected);
+    }
+
+    @Test
+    public void testCovariantEqualsCompactSourceFileMultiple()
+            throws Exception {
+        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        verifyWithInlineConfigParser(
+                getNonCompilablePath(
+                    "InputCovariantEqualsCompactSourceFileMultiple.java"),
+                expected);
+    }
+
+    @Test
+    public void testCovariantEqualsCompactSourceFileNoViolation()
+            throws Exception {
+        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        verifyWithInlineConfigParser(
+                getNonCompilablePath(
+                    "InputCovariantEqualsCompactSourceFileNoViolation.java"),
+                expected);
     }
 
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/covariantequals/InputCovariantEqualsCompactSourceFile.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/covariantequals/InputCovariantEqualsCompactSourceFile.java
@@ -1,0 +1,13 @@
+/*
+CovariantEquals
+
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+
+public boolean equals(String other) {
+    return false;
+}
+
+void main() { }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/covariantequals/InputCovariantEqualsCompactSourceFileMultiple.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/covariantequals/InputCovariantEqualsCompactSourceFileMultiple.java
@@ -1,0 +1,17 @@
+/*
+CovariantEquals
+
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+
+public boolean equals(String other) {
+    return false;
+}
+
+public boolean equals(Integer other) {
+    return false;
+}
+
+void main() { }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/covariantequals/InputCovariantEqualsCompactSourceFileNoViolation.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/covariantequals/InputCovariantEqualsCompactSourceFileNoViolation.java
@@ -1,0 +1,21 @@
+/*
+CovariantEquals
+
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+
+public boolean equals(Object obj) {
+    return true;
+}
+
+public boolean equals(String other) {
+    return false;
+}
+
+public boolean equals(Integer other) {
+    return false;
+}
+
+void main() { }

